### PR TITLE
Calculate `TransactionId` from commitments

### LINKED ISSUE #1439: Calculate TransactionID from TransactionCommitment and OutputCommitment
https://github.com/iotaledger/iota.go/pull/563

### PR DESCRIPTION
**Pull request by DaughterOfMars**
# Description of change

This PR changes the way the transaction hash is calculated, instead using the hash of the first part of the transaction (transaction commitment) plus the merkle root of the outputs (output commitment).

## Links to any relevant issues

Closes #1439

### CODE REVIEW

**Comment by thibault-martinez** [comment]
Rebase or something to get rid of my commits ?

File: sdk/src/types/block/semantic.rs [hunk 1]
```diff
@@ -7,6 +7,7 @@ use core::fmt;
 use hashbrown::{HashMap, HashSet};
 use primitive_types::U256;
 
+use super::payload::signed_transaction::TransactionHash;
```
Comment (thibault-martinez) [inline sdk/src/types/block/semantic.rs]: Merge

Comment (DaughterOfMars) [reply sdk/src/types/block/semantic.rs]: I really wish vscode would stop doing that

Comment (thibault-martinez) [reply sdk/src/types/block/semantic.rs]: Mine doesn't 🤷🏻‍♂️

File: sdk/src/utils/merkle_hasher.rs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        ]
+        .iter()
+        .map(|hash| prefix_hex::decode::<Vec<_>>(hash).unwrap())
+        .collect::<Vec<_>>();
+
+        let hash = MerkleHasher::digest::<Blake2b256>(&hashes).to_vec();
+
+        assert_eq!(
+            prefix_hex::encode(hash),
+            "0xbf67ce7ba23e8c0951b5abaec4f5524360d2c26d971ff226d3359fa70cdb0beb"
```
Comment (thibault-martinez) [inline sdk/src/utils/merkle_hasher.rs]: Can you update the test with these values please https://github.com/iotaledger/iota.go/blob/develop/merklehasher/merkle_hasher_test.go#L19 ?
Just to be sure

Comment (DaughterOfMars) [reply sdk/src/utils/merkle_hasher.rs]: updated

File: sdk/src/types/block/semantic.rs [hunk 3]
```diff
@@ -149,7 +149,7 @@ impl TryFrom<u8> for TransactionFailureReason {
 ///
 pub struct ValidationContext<'a> {
     pub(crate) transaction: &'a Transaction,
-    pub(crate) transaction_hash: [u8; 32],
+    pub(crate) transaction_hash: TransactionHash,
```
Comment (thibault-martinez) [inline sdk/src/types/block/semantic.rs]: I am not fully sure what is being signed actually is the TransactionHash (anymore?). I asked here https://github.com/iotaledger/tips/pull/147/files#r1370380908.

Comment (thibault-martinez) [reply sdk/src/types/block/semantic.rs]: See the answer, as I suspected, this is not the same thing that is signed

Comment (DaughterOfMars) [reply sdk/src/types/block/semantic.rs]: that's quite annoying. So we need two kinds of transaction hash?

Comment (thibault-martinez) [reply sdk/src/types/block/semantic.rs]: If you want it properly typed, yeah. At the very least a new method to get these bytes

Comment (thibault-martinez) [reply sdk/src/types/block/semantic.rs]: TransactionSigningHash? TransactionSignHash?

Comment (DaughterOfMars) [reply sdk/src/types/block/semantic.rs]: I added a new type. I am not sure about all of the places that use the other hash so I added TODOs

File: sdk/src/wallet/account/operations/transaction/sign_transaction.rs [hunk 4]
```diff
@@ -60,7 +60,8 @@ where
                         self.emit(
                             self.details().await.index,
                             WalletEvent::TransactionProgress(TransactionProgressEvent::PreparedTransactionHash(
-                                prefix_hex::encode(prepared_transaction_data.transaction.hash()),
+                                // TODO: Should this be the signing_hash?
```
Comment (thibault-martinez) [inline sdk/src/wallet/account/operations/transaction/sign_transaction.rs]: Yes. Please rename `TransactionProgressEvent::PreparedTransactionHash` to `TransactionProgressEvent::PreparedTransactionSigningHash`, also in the bindings.

File: sdk/src/utils/merkle_hasher.rs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+            [] => Self::empty::<D>(),
+            [leaf] => Self::leaf::<D>(leaf),
+            _ => Self::node::<D>(value),
+        }
+    }
+}
+
+/// Computes the largest power of two less than or equal to `n`.
+fn largest_power_of_two(n: u32) -> usize {
+    debug_assert!(n > 1, "invalid input to `largest_power_of_two`");
```
Comment (thibault-martinez) [inline sdk/src/utils/merkle_hasher.rs:52]: Doesn't really hurt but it's strictly impossible that this happens. Only called in `node`, itself only called with more than one element.

File: sdk/src/types/block/semantic.rs [hunk 6]
```diff
@@ -149,7 +150,7 @@ impl TryFrom<u8> for TransactionFailureReason {
 ///
 pub struct ValidationContext<'a> {
     pub(crate) transaction: &'a Transaction,
-    pub(crate) transaction_hash: [u8; 32],
+    pub(crate) transaction_hash: TransactionSigningHash,
```
Comment (thibault-martinez) [inline sdk/src/types/block/semantic.rs]: I'd also rename the field

File: sdk/src/types/block/semantic.rs [hunk 7]
```diff
@@ -7,6 +7,7 @@ use core::fmt;
 use hashbrown::{HashMap, HashSet};
 use primitive_types::U256;
 
+use super::payload::signed_transaction::TransactionSigningHash;
```
Comment (thibault-martinez) [inline sdk/src/types/block/semantic.rs]: Merge

File: sdk/src/wallet/events/types.rs [hunk 8]
```diff
@@ -263,7 +263,7 @@ impl Serialize for TransactionProgressEvent {
                 kind: 2,
                 event: TransactionProgressEvent_::T2(e),
             },
-            Self::PreparedTransactionHash(e) => TypedTransactionProgressEvent_ {
+            Self::PreparedTransactionSigningHash(e) => TypedTransactionProgressEvent_ {
```
Comment (thibault-martinez) [inline sdk/src/wallet/events/types.rs:267]: Also rename at L232

File: sdk/src/types/block/payload/signed_transaction/mod.rs [hunk 9]
```diff
@@ -46,12 +45,9 @@ impl SignedTransactionPayload {
 
     /// Computes the identifier of a [`SignedTransactionPayload`].
     pub fn id(&self) -> TransactionId {
```
Comment (thibault-martinez) [inline sdk/src/types/block/payload/signed_transaction/mod.rs]: ID is a property of the `Transaction`, I would put it there. I guess it's fine to keep this one and forward though

Comment (DaughterOfMars) [reply sdk/src/types/block/payload/signed_transaction/mod.rs]: moved it moved it

File: bindings/nodejs/lib/types/wallet/event.ts [hunk 10]
```diff
@@ -256,7 +256,7 @@ class PreparedTransactionHashProgress extends TransactionProgress {
      * @param hash The hash of the transaction.
      */
     constructor(hash: HexEncodedString) {
-        super(TransactionProgressType.PreparedTransactionHash);
+        super(TransactionProgressType.PreparedTransactionSigningHash);
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/wallet/event.ts:259]: Rename also L252

File: sdk/src/wallet/events/types.rs [hunk 11]
```diff
@@ -284,7 +284,7 @@ impl<'de> Deserialize<'de> for TransactionProgressEvent {
     fn deserialize<D: serde::Deserializer<'de>>(d: D) -> Result<Self, D::Error> {
         #[derive(Deserialize)]
         struct PreparedTransactionHash_ {
```
Comment (thibault-martinez) [inline sdk/src/wallet/events/types.rs]: Also rename this (sorry I missed it the first time)

File: sdk/src/types/block/payload/signed_transaction/transaction_id.rs [hunk 12]
```diff
@@ -18,3 +18,11 @@ impl TransactionId {
         OutputId::new(self, index)
     }
 }
+
+crate::impl_id!(
+    /// The hash of a transaction commitment and output commitment which is used to create
+    /// [`SignedTransactionPayload`](crate::types::block::payload::SignedTransactionPayload).
```
Comment (thibault-martinez) [inline sdk/src/types/block/payload/signed_transaction/transaction_id.rs]: Wrong comment

File: sdk/src/client/secret/mod.rs [hunk 13]
```diff
@@ -505,7 +505,7 @@ where
     crate::client::Error: From<M::Error>,
 {
     // The transaction_hash gets signed
```
Comment (thibault-martinez) [inline sdk/src/client/secret/mod.rs]: Either update the comment or remove it, but it's useless IMO

File: sdk/src/client/secret/ledger_nano.rs [hunk 14]
```diff
@@ -511,7 +511,7 @@ fn merge_unlocks(
 ) -> Result<Vec<Unlock>, Error> {
     let slot_index = prepared_transaction_data.transaction.creation_slot();
     // The transaction_hash gets signed
```
Comment (thibault-martinez) [inline sdk/src/client/secret/ledger_nano.rs]: same

File: bindings/core/src/method_handler/utils.rs [hunk 15]
```diff
@@ -70,7 +70,8 @@ pub(crate) fn call_utils_method_internal(method: UtilsMethod) -> Result<Response
             Response::TokenId(TokenId::from(foundry_id))
         }
         UtilsMethod::HashTransaction { transaction } => {
-            Response::Hash(prefix_hex::encode(Transaction::try_from_dto(transaction)?.hash()))
+            // TODO: Should this be the signing_hash?
```
Comment (thibault-martinez) [inline bindings/core/src/method_handler/utils.rs]: I don't see why bindings would need the Hash. Let's make it SigningHash for now and we can add Hash if needed. I'd rename the variant then.

Comment (thibault-martinez) [reply bindings/core/src/method_handler/utils.rs]: The method variant, not the response *

File: sdk/src/types/block/payload/signed_transaction/transaction_id.rs (hunk 12 above)
Comment (DaughterOfMars) [reply sdk/src/types/block/payload/signed_transaction/transaction_id.rs]: Wrong in what way?

Comment (thibault-martinez) [reply sdk/src/types/block/payload/signed_transaction/transaction_id.rs]: Signing hash is just the hash of the serialised bytes

File: bindings/core/src/method/utils.rs [hunk 16]
```diff
@@ -124,7 +124,7 @@ pub enum UtilsMethod {
         token_scheme_type: u8,
     },
     /// Computes the hash of a transaction.
```
Comment (thibault-martinez) [inline bindings/core/src/method/utils.rs]:
```suggestion
    /// Computes the signing hash of a transaction.
```

File: sdk/src/types/block/payload/signed_transaction/transaction_id.rs (hunk 12 above)
Comment (DaughterOfMars) [reply sdk/src/types/block/payload/signed_transaction/transaction_id.rs]: oh I got it backwards oops

File: sdk/src/wallet/events/types.rs [hunk 17]
```diff
@@ -302,12 +302,12 @@ impl<'de> Deserialize<'de> for TransactionProgressEvent {
                 2 => Self::PreparedTransaction(Box::new(PreparedTransactionDataDto::deserialize(value).map_err(
                     |e| serde::de::Error::custom(format!("cannot deserialize PreparedTransactionDataDto: {e}")),
                 )?)),
-                3 => Self::PreparedTransactionHash(
-                    PreparedTransactionHash_::deserialize(value)
+                3 => Self::PreparedTransactionSigningHash(
+                    PreparedTransactionSigningHash_::deserialize(value)
                         .map_err(|e| {
                             serde::de::Error::custom(format!("cannot deserialize PreparedTransactionHash: {e}"))
```
Comment (thibault-martinez) [inline sdk/src/wallet/events/types.rs]:
```suggestion
                            serde::de::Error::custom(format!("cannot deserialize PreparedTransactionSigningHash: {e}"))
```

File: sdk/src/client/secret/ledger_nano.rs [hunk 18]
```diff
@@ -271,7 +271,7 @@ impl SecretManage for LedgerSecretManager {
 
         // pack transaction and hash into vec
```
Comment (thibault-martinez) [inline sdk/src/client/secret/ledger_nano.rs]:
```suggestion
        // pack transaction and signing hash into vec
```
Or remove

File: sdk/src/client/secret/ledger_nano.rs [hunk 19]
```diff
@@ -289,9 +289,9 @@ impl SecretManage for LedgerSecretManager {
         if blind_signing {
             // prepare signing
             log::debug!("[LEDGER] prepare_blind_signing");
-            log::debug!("[LEDGER] {:?} {:?}", input_bip32_indices, transaction_hash);
+            log::debug!("[LEDGER] {:?} {:?}", input_bip32_indices, transaction_signing_hash);
             ledger
-                .prepare_blind_signing(input_bip32_indices, transaction_hash)
+                .prepare_blind_signing(input_bip32_indices, transaction_signing_hash)
```
Comment (thibault-martinez) [inline sdk/src/client/secret/ledger_nano.rs:294]: If you could rename L437 that would be nice 🙏🏻

File: sdk/src/wallet/events/types.rs [hunk 20]
```diff
@@ -216,7 +216,7 @@ pub enum TransactionProgressEvent {
     /// Prepared transaction.
     PreparedTransaction(Box<PreparedTransactionDataDto>),
     /// Prepared transaction hash hex encoded, required for blindsigning with a ledger nano
```
Comment (thibault-martinez) [inline sdk/src/wallet/events/types.rs]:
```suggestion
    /// Prepared transaction signing hash hex encoded, required for blindsigning with a ledger nano
```

File: sdk/src/utils/merkle_hasher.rs (hunk 5 above)
Comment (thibault-martinez) [reply sdk/src/utils/merkle_hasher.rs:52]: Resolve if you won't change

File: sdk/src/types/block/payload/signed_transaction/transaction_id.rs [hunk 21]
```diff
@@ -18,3 +19,10 @@ impl TransactionId {
         OutputId::new(self, index)
     }
 }
+
+crate::impl_id!(
+    /// The hash of a [`Transaction`](crate::types::block::payload::signed_transaction::Transaction).
```
Comment (thibault-martinez) [inline sdk/src/types/block/payload/signed_transaction/transaction_id.rs]:
```suggestion
    /// The signing hash of a [`Transaction`](crate::types::block::payload::signed_transaction::Transaction).
```

**Review by Thoralf-M — DISMISSED** [review]

**Review by Thoralf-M — DISMISSED** [review]

File: bindings/nodejs/lib/types/wallet/event.ts [hunk 22]
```diff
@@ -249,14 +249,14 @@ class PreparedTransactionProgress extends TransactionProgress {
 /**
  * A 'prepared transaction hash' progress.
  */
-class PreparedTransactionHashProgress extends TransactionProgress {
+class PreparedTransactionSigningHashProgress extends TransactionProgress {
     hash: HexEncodedString;
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/wallet/event.ts]: I think you have renamed the field in rust so it needs to be the same to be parsed correctly

Comment (DaughterOfMars) [reply bindings/nodejs/lib/types/wallet/event.ts]: done

File: bindings/nodejs/lib/types/wallet/event.ts [hunk 23]
```diff
@@ -250,14 +250,14 @@ class PreparedTransactionProgress extends TransactionProgress {
  * A 'prepared transaction hash' progress.
  */
 class PreparedTransactionSigningHashProgress extends TransactionProgress {
-    hash: HexEncodedString;
+    signing_hash: HexEncodedString;
```
Comment (Thoralf-M) [inline bindings/nodejs/lib/types/wallet/event.ts]: No camelCase? (also needs to be done on the Rust side then)

**Review by thibault-martinez — DISMISSED** [review]

File: bindings/nodejs/lib/types/wallet/event.ts (hunk 23 above)
Comment (thibault-martinez) [reply bindings/nodejs/lib/types/wallet/event.ts]: Ah yeah probably need

**Review by Thoralf-M — DISMISSED** [review]

File: bindings/nodejs/lib/types/wallet/event.ts [hunk 24]
```diff
@@ -249,15 +249,15 @@ class PreparedTransactionProgress extends TransactionProgress {
...
  * A 'prepared transaction hash' progress.
  */
-class PreparedTransactionHashProgress extends TransactionProgress {
-    hash: HexEncodedString;
+class PreparedTransactionSigningHashProgress extends TransactionProgress {
+    signingHash: HexEncodedString;
 
     /**
-     * @param hash The hash of the transaction.
+     * @param signing_hash The signing hash of the transaction.
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/wallet/event.ts]:
```suggestion
     * @param signingHash The signing hash of the transaction.
```

**Review by thibault-martinez — APPROVED** [review]

**Review by Thoralf-M — APPROVED** [review]

**Review by Alex6323 — APPROVED** [review]

**Review by kwek20 — APPROVED** [review]

**Pull request merged**